From ddf2b73b99b9c64ac6fdd7ed50cb320473384141 Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Mon, 10 Apr 2023 10:27:52 +0800
Subject: [PATCH] xen/arm: Allow MPAM v1.1 in MSC hardware probing process

According to MPAM specification (ARM DDI 0598C.a), there are different
versions of MPAM to be probed, and different versions of MPAM can
be identified by the MPAMF_AIDR register from the combination of
fields {ArchMajorRev, ArchMinorRev}.

Since hardware with MPAM v1.1 is available, allow such hardware to use
the MPAM driver during the hardware probing process.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-6012
---
 xen/arch/arm/include/asm/mpam.h | 3 ++-
 xen/arch/arm/mpam.c             | 5 +++--
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/xen/arch/arm/include/asm/mpam.h b/xen/arch/arm/include/asm/mpam.h
index b72179296f70..6dc7b7120f4c 100644
--- a/xen/arch/arm/include/asm/mpam.h
+++ b/xen/arch/arm/include/asm/mpam.h
@@ -16,7 +16,8 @@
  * Arm Architecture Reference Manual Supplement - Memory System Resource
  * Partitioning and Monitoring (MPAM), for Armv8-A. DDI 0598A.a
  */
-#define MPAM_ARCHITECTURE_V1    0x10
+#define MPAM_ARCHITECTURE_V1_0    0x10
+#define MPAM_ARCHITECTURE_V1_1    0x11
 
 /* Memory mapped control pages: */
 /* ID Register offsets in the memory mapped page */
diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index fba4b93bd7c9..e8212f438081 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -893,9 +893,10 @@ static int __init mpam_msc_hw_probe(struct mpam_msc *msc)
     BUG_ON(!spin_is_locked(&msc->lock));
 
     spin_lock_irqsave(&msc->hw_lock, flags);
-    if ( mpam_read_reg(msc, MPAMF_AIDR) != MPAM_ARCHITECTURE_V1 )
+    if ( mpam_read_reg(msc, MPAMF_AIDR) != MPAM_ARCHITECTURE_V1_0 &&
+         mpam_read_reg(msc, MPAMF_AIDR) != MPAM_ARCHITECTURE_V1_1 )
     {
-        printk("MPAM: MSC%d does not match MPAM architecture v1.0\n",
+        printk("MPAM: MSC%d does not match MPAM architecture v1.0 nor v1.1\n",
                msc->identifier);
         spin_unlock_irqrestore(&msc->hw_lock, flags);
         return -EIO;
-- 
2.34.1

